client/scala/src/main/scala/ai/verta/blobs/dataset/Dataset.scala

```diff
@@ -5,14 +5,110 @@ import ai.verta.blobs._
 
 import java.security.{MessageDigest, DigestInputStream}
 import java.io.{File, FileInputStream}
+import java.nio.file.Paths
 
 import scala.collection.mutable.HashMap
 import scala.util.{Failure, Success, Try}
+import scala.concurrent.ExecutionContext
 
 trait Dataset extends Blob {
   protected val contents: HashMap[String, FileMetadata] // for deduplication and comparing
   private[verta] val enableMDBVersioning: Boolean // whether to version the blob with ModelDB
 
+  // mutable state, populated when getting blob from commit
+  /** TODO: Figure out a way to remove this */
+  // Function to downwload a component, given its path in the blob, the blob's path in the commit
+  // and the file pointing to local path to download to:
+  private[verta] var downloadFunction: Option[(String, String, File) => Try[Unit]] = None
+  private[verta] var blobPath: Option[String] = None // path to the blob in the commit
+
+  /** Downloads componentPath from this dataset if ModelDB-managed versioning was enabled
+   *  Currently, only support downloading to a specific path
+   *  @param componentPath Original path of the file or directory in this dataset to download
+   *  @param downloadToPath Path to download to
+   *  @return Whether the download attempts succeed.
+   */
+  def download(
+    componentPath: Option[String] = None,
+    downloadToPath: String
+  )(implicit ec: ExecutionContext): Try[Unit] = {
+    /** TODO: Make downloadToPath optional */
+
+    if (!enableMDBVersioning)
+      Failure(new IllegalStateException("This blob did not allow for versioning"))
+    else if (downloadFunction.isEmpty || blobPath.isEmpty)
+      Failure(new IllegalStateException(
+        "This dataset cannot be used for downloads. Consider using `commit.get()` to obtain a download-capable dataset"
```

Comment:
Under which scenarios would this happen?

Reply:
So currently the way Python client work (which I'm trying to emulate) is that:
```scala
val originalBlob = PathBlob("some-path", true).get
val newCommit = commit.update("path-blob", originalBlob).flatMap(_.save("some-msg")).get

val downloadingBlob = newCommit.get("path-blob").get
downloadingBlob.download()

originalBlob.download() // this would fail
```
So only a blob obtained via commit.get can download stuff

Reply:
Added a small test to demonstrate the expected behavior (i.e failures).